consensus/polybft/proposer_calculator.go

```diff
@@ -0,0 +1,355 @@
+	// It needs to be sufficiently small to, in all cases:
+	// 1. prevent clipping in incrementProposerPriority()
+	// 2. let (diff+diffMax-1) not overflow in IncrementProposerPriority()
+	// (Proof of 1 is tricky, left to the reader).
+	// It could be higher, but this is sufficiently large for our purposes,
+	// and leaves room for defensive purposes.
+	maxTotalVotingPower = int64(math.MaxInt64) / 8
+
+	// priorityWindowSizeFactor - is a constant that when multiplied with the
+	// total voting power gives the maximum allowed distance between validator
+	// priorities.
+	priorityWindowSizeFactor = 2
+)
+
+var (
+	// errInvalidTotalVotingPower is returned if the total voting power is zero
+	errInvalidTotalVotingPower = errors.New(
+		"invalid voting power configuration provided: total voting power must be greater than 0")
+)
+
+// ProposerCalculator interface of the current validator set
+type ProposerCalculator interface {
+	// CalcProposer calculates next proposer based on the passed round
+	CalcProposer(round uint64) (types.Address, error)
+
+	// IncrementProposerPriority increments priorities number of times
+	IncrementProposerPriority(times uint64) error
+
+	// GetLatestProposer returns latest calculated proposer
+	GetLatestProposer() (types.Address, bool)
+}
+
+type validatorCalcMetadata struct {
+	Metadata         *ValidatorMetadata
+	ProposerPriority int64
+}
+
+func (a validatorCalcMetadata) IsBetter(b *validatorCalcMetadata) bool {
+	if b == nil || a.ProposerPriority > b.ProposerPriority {
+		return true
+	} else if a.ProposerPriority == b.ProposerPriority {
+		return bytes.Compare(a.Metadata.Address.Bytes(), b.Metadata.Address.Bytes()) <= 0
+	} else {
+		return false
+	}
+}
+
+// newValidatorCalcMetadata returns a new validator with the given pubkey and voting power.
+func newValidatorCalcMetadata(metadata *ValidatorMetadata, priority int64) *validatorCalcMetadata {
+	return &validatorCalcMetadata{
+		Metadata:         metadata,
+		ProposerPriority: priority,
+	}
+}
+
+type proposerCalculator struct {
+	// validators represents current list of validators with their priority
+	validators []*validatorCalcMetadata
+
+	// proposer of a block
+	proposer *validatorCalcMetadata
+
+	// current proposer is calculated in this round - optimization
+	round uint64
+
+	// total voting power
+	totalVotingPower int64
+
+	// rw mutex
+	lock *sync.RWMutex
+
+	// logger instance
+	logger hclog.Logger
+}
+
+// NewProposerCalculator creates a new proposer calculator object.
+func NewProposerCalculator(valz AccountSet, totalVotingPower int64, logger hclog.Logger) (*proposerCalculator, error) {
+	var validators = make([]*validatorCalcMetadata, len(valz))
+	for i, v := range valz {
+		validators[i] = newValidatorCalcMetadata(v, 0)
+	}
+
+	proposerCalc := &proposerCalculator{
+		totalVotingPower: totalVotingPower,
+		validators:       validators,
+		lock:             &sync.RWMutex{},
+		round:            uint64(math.MaxUint64),
```

Comment:
Just curious, why are we setting the round to `MaxUint64` initially? Shouldn't it just be 0?

Reply:
If its 0, than the first `CalcProposer` call will return "cached" value

Reply:
Will work with zero too. Changed that with https://github.com/0xPolygon/polygon-edge/pull/984/commits/4050ffe6ebbc88e2b3fa15075acc343e14071771